storage/zynq7000-flash/qspi.c

```diff
@@ -0,0 +1,436 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Quad-SPI Controller driver
+ *
+ * Copyright 2021, 2022 Phoenix Systems
+ * Author: Hubert Buczynski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include "qspi.h"
+
+#include <errno.h>
+#include <string.h>
+#include <sys/time.h>
+#include <sys/mman.h>
+#include <sys/threads.h>
+#include <sys/platform.h>
+#include <sys/interrupt.h>
+
+#include <phoenix/arch/zynq7000.h>
+
+
+enum { cr = 0, sr, ier, idr, imr, er, dr, txd00, rxd, sicr, txth, rxth, gpio,
+	   lpbk = 0xe, txd01 = 0x20, txd10, txd11,
+	   lqspi_cr = 0x28, lqspi_sr, modid = 0x3f };
+
+
+struct {
+	volatile uint32_t *base;
+	handle_t cond;
+	handle_t inth;
+	handle_t lock;
+} qspi_common;
+
+
+static inline void qspi_dataMemoryBarrier(void)
+{
+	__asm__ volatile ("dmb");
+}
+
+
+static int qspi_irqHandler(unsigned int n, void *arg)
+{
+	*(qspi_common.base + idr) |= (1 << 2);
+
+	return 1;
+}
+
+
+static inline time_t qspi_timeMsGet(void)
+{
+	struct timeval time;
+	gettimeofday(&time, NULL);
+
+	return (time_t)((time.tv_sec * 1000LL) + (time.tv_usec / 1000));
+}
+
+
+void qspi_stop(void)
+{
+	/* Clean up RX Fifo */
+	while ((*(qspi_common.base + sr) & (1 << 4)))
+		*(qspi_common.base + rxd);
+
+	*(qspi_common.base + cr) |= (1 << 10);
```

Comment:
This line looks like something that maybe need qspi_dataMemoryBarrier

Reply:
Done